### PR TITLE
Separate matrix graph construction from linear system construction

### PR DESCRIPTION
**Pull request by jhux2**
**Pull-request type:**
- [ ] Bug fix 
- [ ] Documentation update
- [x] Feature enhancement

## Description of the pull-request

Addresses #456.

Factor out matrix graph construction from linear system construction.  The ultimate goal is to be able to reuse the matrix graph across multiple linear system construction phases.

## Checklist

**NOTE** The checklist below is a suggestion and not mandatory. You don't have
to _check all boxes_ to submit a pull request. However, please add a brief
explanation in your pull request summary explaining the omission for the benefit
of reviewers and developers.

*All pull requests*
- [x ] Builds successfully (*must test on at least one system/compiler combination*)
  - Operating systems 
    - [x ] Linux
    - [ ] MacOS
  - Compilers 
    - [ x] GCC
    - [ ] LLVM/Clang
    - [ ] Intel compilers
    - [ ] NVIDIA CUDA
- [ ] Compiles without warnings
- [ ] Passes all unit tests
- [ ] Passes all regression tests
- [ ] Documentation builds without errors

*For new code updates*
- [ ] Documentation updates - additions to user, theory, and verification manuals
- [ ] New unit tests providing coverage for new code, bug fixes etc.
- [ ] New regression tests exercising the new code

---
This change is [Reviewable](https://reviewable.io/reviews/exawind/nalu-wind/473)

### CODE REVIEW

**Comment by jhux2** [comment]
@sayerhs @lucbv @rcknaus

**Comment by jhux2** [comment]
> I would suggest that the functions/methods copy_stk_to_tpetra and copy_tpetra_to_stk don't go in the CrsGraph class.

@alanw0 I agree.  I left both those methods in the TpetraLinearSystem/TpetraSegregatedLinearSystem classes.

File: include/CrsGraphTypes.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+class ParameterList;
+
+}
+
+namespace sierra{
+namespace nalu{
+
+namespace GraphTypes {
+
+typedef long   GlobalOrdinal; // MUST be signed
```
[the remark this answers is not in the document]
Comment (jhux2) [reply include/CrsGraphTypes.h]: @lucbv Thanks, I've fixed the types and resolved some previous merge conflicts.

**Comment by jhux2** [comment]
Currently, all tests pass except the following.
```
         18 - dgNonConformalThreeBlade (Failed)
	 37 - hoHelium (Failed)
	 38 - hoVortex (Failed)
	 63 - steadyTaylorVortex (Failed)
	 64 - SSTChannelEdge (Failed)
	 65 - tamsChannelEdge (Failed)
	 66 - SSTTAMSChannelEdge (Failed)
	 67 - tamsChannelElem (Failed)
	 68 - SSTTAMSOversetRotCylinder (Failed)
	 71 - airfoilRANSEdge (Failed)
	 72 - airfoilRANSElem (Failed)
	 73 - dgncThreeBladeHypre (Failed)

```

Tests 64, 65, 66, 67, 68, 71, 72, and 73 all require hypre, which I don't have enabled.  So I'm not worried about those failures.

Tests 18, 37, 38, and 63 have the following sort of error.

```
18: PP2 lid=-1 for rowGid=1 lid=-1 for rowGid=25968 sent from proc 3
18: P2 lid=-1 for rowGid=26234 sent from proc 3
18: P2 lid=-1 for rowGid=26417 sent from proc 3
18: P2 lid=-112985 sent from proc 0
18: P1 lid=-1 for rowGid=12986 sent from proc 0 for rowGid=26580 sent from proc 3
18: P2 lid=-1 for rowGid=26234 sent from proc 3
18: P2 lid=-1 for rowGid=26514 sent from proc 3
18: P2
18: P1 lid=-1 for rowGid=12987 sent from proc 0
18:  lid=-1 for rowGid=26580 sent from proc 3
18: P2 lid=-1 for rowGid=26234 sent from proc 3
18: P2 lid=-1 for rowGid=26581 sent from proc 3P1 lid=-1 for rowGid=12988 sent from proc 0
```

**Comment by lucbv** [comment]
@jhux if some tests require hypre then we should modify the CTest logic and guard them with 
```
if(ENABLE_HYPRE)

endif()
```
I will make a PR to clean that up.

**Comment by jhux2** [comment]
@sayerhs @rcknaus Could the failures in 18, 37, 38, and 63 be due to input deck BCs that need to be modified?

**Comment by sayerhs** [comment]
@jhux2 37, 38, and 63 are higher order elements, so I suspect we are missing something in the graph. `steadyTaylorVortex` would be the place to start on figuring that out. 

18 is a moving mesh simulation with a non-conformal interface. So the interior algorithm will not account for the additional graph nodes coming from the boundary elements from the _other_ mesh across the non-conformal interface.

**Comment by rcknaus** [comment]
@jhux2 I don't think so.  `steadyTaylorVortex` has all periodic conditions, so the graph should be the same.  

I looked at the higher order cases a little bit. It looks like the issue is that the higher order cases all solve an extra "system" equation (numDof =3).  My guess is some of the numDof logic is getting done twice.

File: include/Realm.h [hunk 2]
```diff
@@ -438,6 +440,11 @@ class Realm {
   
   EquationSystems equationSystems_;
 
+  //single-DOF graph
+  Teuchos::RCP<CrsGraph> scalarGraph_;
+  //multiple-DOF graph
+  Teuchos::RCP<CrsGraph> systemGraph_;
+
```
Comment (rcknaus) [inline include/Realm.h]: We have one instance of a linear system a "tensor" for the dof, so 4 components in 2D or 9 in 3D.

File: src/CrsGraph.C [hunk 3]
```diff
@@ -0,0 +1,1031 @@
...
+  LocalOrdinal localId = 0;
+  int gstart = 0;
+
+  GlobalOrdinal gomin   = ownedRowsMap_->getMinGlobalIndex();
+
+  for(stk::mesh::Entity entity : owned_nodes) {
+    const stk::mesh::EntityId entityId = *stk::mesh::field_data(*realm_.naluGlobalId_, entity);
+    myLIDs_[entityId] = numDof_*localId++;
+    auto *  thisgid = stk::mesh::field_data(*realm_.tpetGlobalId_, entity);
+    auto basegid = gomin + numDof_ * gstart;
```
Comment (rcknaus) [inline src/CrsGraph.C]: The dependency on `numDof_` here means that the tpetra global id field can't be used in the `CrsGraph::finalizeGraph()` step unless the last call to a `CrsGraph::beginConstruction()` method used the same `numDof_` parameter.

This I think is the issue with the failing tests.  We could bypass this issue though for now by making sure that a `CrsGraph` only ever calls `CrsGraph::finalizeGraph()` once.

File: src/CrsGraph.C [hunk 4]
```diff
@@ -0,0 +1,1031 @@
...
+          ownersAndGids_.insert(std::make_pair(bulkData.parallel_owner_rank(master), gid));
+        }
+      }
+    }
+  }
+} //storeOwnersForShared
+
+void CrsGraph::finalizeGraph()
+{
+  ThrowRequire(inConstruction_);
```
Comment (rcknaus) [inline src/CrsGraph.C]: We probably want to avoid "re-finalizing" the same graph.  Combined with the issue with the tpetra global id potentially being in a bad state I mentioned before, finalizations after the first call may be incorrect depending on the order in which the equation systems are sequenced.

File: src/AssembleElemSolverAlgorithm.C [hunk 5]
```diff
@@ -74,10 +74,8 @@ AssembleElemSolverAlgorithm::AssembleElemSolverAlgorithm(
 void
 AssembleElemSolverAlgorithm::initialize_connectivity()
 {
-  if (entityRank_ == stk::topology::ELEM_RANK)
-    eqSystem_->linsys_->buildElemToNodeGraph(partVec_);
-  else if (entityRank_ == realm_.meta_data().side_rank())
-    eqSystem_->linsys_->buildFaceToNodeGraph(partVec_);
+  eqSystem_->linsys_->buildFaceElemToNodeGraph(partVec_);
```
Comment (sayerhs) [inline src/AssembleElemSolverAlgorithm.C]: Changing `initialize_connectivity` is not what I had in mind when I suggested that we uniformly call `buildElemToNodeGraph` etc.

**Comment by jhux2** [comment]
Here are the remaining test failures:
```
6: FAIL: airfoilRANSEdgeTrilinos.................     4.4961s 1.4686e-01 2.2813e-07
22: FAIL: ductWedge...............................     0.6109s 5.2951e-11 4.2101e-07
23: FAIL: edgeHybridFluids........................    34.2733s 3.1472e-14 3.5561e-11
24: FAIL: edgePipeCHT.............................     4.9603s 1.8287e-12 2.3133e-12
25: FAIL: ekmanSpiral.............................    11.2711s 9.8148e-07 8.9743e-05
43: FAIL: karmanVortex............................     0.4997s 2.5751e-08 9.4357e-07
50: FAIL: nonIsoEdgeOpenJet.......................    12.9589s 5.6479e-11 1.5776e-07
53: FAIL: nonIsoNonUniformEdgeOpenJet.............    34.3867s 4.0505e-08 2.2234e-06
```

File: src/LowMachEquationSystem.C [hunk 6]
```diff
@@ -2480,8 +2480,13 @@ MomentumEquationSystem::register_overset_bc()
 void
 MomentumEquationSystem::initialize()
 {
+  for (auto it = equationSystems_.equationSystemVector_.begin(); it != equationSystems_.equationSystemVector_.end(); ++it) {
+    //std::cout << "MomentumEquationSystem::initialize() : equationSystem name = " <<  (*it)->name_ << "," << (*it)->convergenceTolerance_ << std::endl;
+    if ((*it)->name_ == "LowMachEOSWrap")
+      convergenceTolerance_ = (*it)->convergenceTolerance_;
+  }
```
Comment (sayerhs) [inline src/LowMachEquationSystem.C]: Why is this here? The `convergenceTolerance_` values are set here 

https://github.com/Exawind/nalu-wind/blob/b9d2834c770be4786e23f1d32e28f718af82c991/src/LowMachEquationSystem.C#L320-L326

File: src/LowMachEquationSystem.C [hunk 7]
```diff
@@ -3683,6 +3688,11 @@ ContinuityEquationSystem::register_overset_bc()
 void
 ContinuityEquationSystem::initialize()
 {
+  for (auto it = equationSystems_.equationSystemVector_.begin(); it != equationSystems_.equationSystemVector_.end(); ++it) {
+    //std::cout << "MomentumEquationSystem::initialize() : equationSystem name = " <<  (*it)->name_ << "," << (*it)->convergenceTolerance_ << std::endl;
+    if ((*it)->name_ == "LowMachEOSWrap")
+      convergenceTolerance_ = (*it)->convergenceTolerance_;
+  }
```
Comment (sayerhs) [inline src/LowMachEquationSystem.C]: See comment in `MomentumEquationSystem` above

File: src/LowMachEquationSystem.C [hunk 8]
```diff
@@ -3706,7 +3716,7 @@ ContinuityEquationSystem::initialize()
   }
 
   solverAlgDriver_->initialize_connectivity();
-  linsys_->finalizeLinearSystem();
+  //linsys_->finalizeLinearSystem();
```
Comment (sayerhs) [inline src/LowMachEquationSystem.C]: Cleanup?

File: src/LowMachEquationSystem.C (hunk 6 above)
Comment (jhux2) [reply src/LowMachEquationSystem.C]: @sayerhs With the refactor, this code is no longer being called, hence momentum and continuity are defaulting to a NL tolerance of 1.

Comment (jhux2) [reply src/LowMachEquationSystem.C:2487]: Here are the backtraces that show where the flow of control diverges:

Nalu-Wind master

```
#1  0x000000000064972f in sierra::nalu::EquationSystems::initialize (this=this@entry=0x141ec30)
    at /home/jhu/exawind/sources/nalu-wind-exawind/src/EquationSystems.C:713
713     eqSys->initialize();
(gdb)
#2  0x000000000053ab19 in sierra::nalu::Realm::initialize (this=0x141e970) at /home/jhu/exawind/sources/nalu-wind-exawind/src/Realm.C:566
566   equationSystems_.initialize();
(gdb)
#3  0x0000000000548626 in sierra::nalu::Realms::initialize (this=0x144eb00) at /home/jhu/exawind/sources/nalu-wind-exawind/src/Realms.C:78
78      realmVector_[irealm]->initialize();
(gdb)
#4  0x0000000000513d4d in sierra::nalu::Simulation::initialize (this=this@entry=0x7fffffffa750)
    at /home/jhu/exawind/sources/nalu-wind-exawind/src/Simulation.C:146
146   realms_->initialize();
(gdb)
#5  0x000000000045cf01 in main (argc=<optimized out>, argv=<optimized out>) at /home/jhu/exawind/sources/nalu-wind-exawind/nalu.C:168
168   sim.initialize();
```

CrsGraph branch

```
    at /home/jhu/exawind/sources/nalu-wind-jhux2/src/LowMachEquationSystem.C:3687
3687	{
(gdb) n
3688	  for (auto it = equationSystems_.equationSystemVector_.begin(); it != equationSystems_.equationSystemVector_.end(); ++it) {
(gdb) up
#1  0x000000000064dd8f in sierra::nalu::EquationSystems::<lambda(sierra::nalu::EquationSystem*)>::operator() (
    __closure=<optimized out>, eqsys=0x16576d0) at /home/jhu/exawind/sources/nalu-wind-jhux2/src/EquationSystems.C:776
776	      eqsys->initialize();
(gdb)
#2  sierra::nalu::(anonymous namespace)::initialize_connectivity<sierra::nalu::EquationSystems::initialize()::<lambda(sierra::nalu::EquationSystem*)> > (eqSysVec=std::vector of length 4, capacity 4 = {...}, realm=..., func=...)
    at /home/jhu/exawind/sources/nalu-wind-jhux2/src/EquationSystems.C:97
97	      func(eqsys);
(gdb)
#3  sierra::nalu::EquationSystems::initialize (this=this@entry=0x1421350)
    at /home/jhu/exawind/sources/nalu-wind-jhux2/src/EquationSystems.C:772
772	  initialize_connectivity(
(gdb)
#4  0x000000000053c769 in sierra::nalu::Realm::initialize (this=0x1421090)
    at /home/jhu/exawind/sources/nalu-wind-jhux2/src/Realm.C:566
566	  equationSystems_.initialize();
(gdb)
#5  0x000000000054a2d6 in sierra::nalu::Realms::initialize (this=0x1451640)
    at /home/jhu/exawind/sources/nalu-wind-jhux2/src/Realms.C:78
78	    realmVector_[irealm]->initialize();
(gdb)
#6  0x000000000051593d in sierra::nalu::Simulation::initialize (this=this@entry=0x7fffffffa750)
    at /home/jhu/exawind/sources/nalu-wind-jhux2/src/Simulation.C:146
146	  realms_->initialize();
```

**Comment by sayerhs** [comment]
Thanks @jhux For the trace. I am sorry I didn't get a chance to look into this. I'd like to see if there's a good way to remove this regression. There's another equation system SST that's a wrapper and I want to make sure that it is behaving correctly and it wasn't just that our regression test suite didn't catch it.

**Comment by jhux2** [comment]
> Thanks @jhux For the trace. I am sorry I didn't get a chance to look into this. I'd like to see if there's a good way to remove this regression. There's another equation system SST that's a wrapper and I want to make sure that it is behaving correctly and it wasn't just that our regression test suite didn't catch it.

No worries.  I didn't quite understand all the flow control in the system setup, so I'll be the first to admit that my change is a little hacky.

File: src/EquationSystems.C [hunk 9]
```diff
@@ -46,6 +50,71 @@
...
+  std::vector<EquationSystem*>& eqSysVec,
+  LambdaFunc func)
+{
+  std::map<int, std::vector<EquationSystem*>> eqSysMap;
+  // Group equation systems based on numDof. Note that this is safe for
+  // reinitialization also as the `linsys_` instance is still valid and does not
+  // get reset until the call to `EquationSystem::reinitialize_linear_system`
+  for (auto* eqsys: eqSysVec) {
+    // Skip wrapper type equation systems that don't have a LinearSystem instance
+    if (eqsys->linsys_ == nullptr) continue;
```
Comment (sayerhs) [inline src/EquationSystems.C:82]: @jhux2 This is where the regression was introduced. We should change this line to 

```C++

// For "wrapper" equation systems, execute their default initialize() method to allow convergence tolerance 
// to be passed to their children
if (eqsys->linsys_ == nullptr)
  eqsys->initialize();
```

This will allow you to remove the stuff had to add in `momentum` and `continuity`

Comment (sayerhs) [reply src/EquationSystems.C:82]: This change will allow us to also handle `SST` wrapper 

https://github.com/Exawind/nalu-wind/blob/b9d2834c770be4786e23f1d32e28f718af82c991/src/ShearStressTransportEquationSystem.C#L110-L116

Comment (jhux2) [reply src/EquationSystems.C]: @sayerhs Thanks, that fixed it!  We're just left with just a few failures, which seem benign.

Is this ok to merge?

```
6: FAIL: airfoilRANSEdgeTrilinos.................     4.7904s 1.4686e-01 2.2813e-07
22: FAIL: ductWedge...............................     0.6451s 5.2951e-11 4.2101e-07
23: FAIL: edgeHybridFluids........................    35.4491s 3.1472e-14 3.5561e-11
24: FAIL: edgePipeCHT.............................     5.6124s 1.8287e-12 2.3133e-12
25: FAIL: ekmanSpiral.............................    12.8755s 9.8148e-07 8.9743e-05
43: FAIL: karmanVortex............................     0.5497s 2.5751e-08 9.4357e-07
50: FAIL: nonIsoEdgeOpenJet.......................    15.8953s 5.6479e-11 1.5776e-07
53: FAIL: nonIsoNonUniformEdgeOpenJet.............    43.9456s 4.0505e-08 2.2234e-06
```
